django_lightweight_queue/runner.py

```diff
@@ -110,51 +88,33 @@ def handle_term(signum, stack):
                 workers[(queue, worker_num)] = worker
                 worker.start()
 
-        while True:
-            try:
-                log.debug("Checking back channel for items")
-
-                # We don't use the timeout kwarg so that when we get a TERM
-                # signal we don't have problems with interrupted system calls.
-                msg = back_channel.get_nowait()
-
-                queue, worker_num, timeout, sigkill_on_stop = msg
-            except (Empty, EOFError):
-                break
-
-            worker = workers[(queue, worker_num)]
-
-            kill_after = None
-            if timeout is not None:
-                kill_after = time.time() + timeout
+        time.sleep(1)
 
-            log.debug(
-                "Setting kill_after=%r and sigkill_on_stop=%r for %s",
-                kill_after,
-                sigkill_on_stop,
-                worker.name,
-            )
-            worker.kill_after = kill_after
-            worker.sigkill_on_stop = sigkill_on_stop
+    # The cron scheduler is always safe to kill.
+    os.kill(cron_scheduler.pid, signal.SIGKILL)
+    cron_scheduler.join()
 
-        time.sleep(1)
+    def signal_workers(signum, condition):
+        for worker in workers.values():
+            if worker is None:
+                continue
 
-    # Filter out workers which might not have yet been started
-    alive_workers = [x for x in workers.values() if x is not None]
+            if not condition(worker):
+                continue
 
-    # Kill all the killable workers. While we could do this second and thus give
-    # these workers a bit more time while we wait for the others, the extra time
-    # these workers would get is likely small and not worth the trade-off that
-    # the master could be killed while we're trying to tidy up.
-    for worker in alive_workers:
-        if worker.sigkill_on_stop:
-            log.info("Sending SIGKILL to %s", worker.name)
             try:
-                os.kill(worker.pid, signal.SIGKILL)
+                os.kill(worker.pid, signum)
             except OSError:
                 pass
 
-    for worker in alive_workers:
+    # SIGUSR2 all the workers. This sets a flag asking them to shut down
+    # gracefully, or kills them immediately if they are receptive to that
+    # sort of abuse.
+    signal_workers(signal.SIGUSR2, lambda worker: True)
```

Comment:
I think `condition` here is unused, so we could remove it?

Reply:
ee32d86